test/libyul/yulOptimizerTests/fullSuite/medium.yul

```diff
@@ -23,7 +23,14 @@
 //
 // {
 //     {
-//         sstore(0, add(mload(0x40), 128))
-//         sstore(1, 0x20)
+//         let p := mload(0x40)
+//         mstore(0x40, add(p, 96))
+//         let _1 := add(p, 128)
+//         mstore(_1, 2)
+//         mstore(0x40, 0x20)
+//         for { } iszero(1) { }
+//         { }
+//         sstore(0, _1)
+//         sstore(1, mload(0x40))
```

Comment:
This strikes as quite a bad regression - not that the case wasn't fully artificial to begin with, but the test case doesn't really make sense with this result anymore...

And it is rather suspicious that suddenly none of the things like removing empty for loops or resolving memory accesses happen anymore here :-). So the question is: can you easily explain why that is?

Reply:
(The solution here may rather be changing the test case, rather than trying to tweak the sequence further, but it'd still be interesting to have an explanation for why this changes this drastically)

Reply:
Or well, maybe we don't even have to change the test case - we can still revisit this, if we're getting actual reports about real-world cases.

Reply:
I investigated this regression for the previous sequence candidate: https://github.com/ethereum/solidity/pull/14928#pullrequestreview-1964569199. I see that now the regression is a bit worse - that `for` loop wasn't there back then - but the same tweak should still resolve it here.

I ultimately decided not to include that tweak though, because when I benchmarked it, it seemed to make both bytecode size and runtime gas slightly worse: https://github.com/ethereum/solidity/pull/14928#discussion_r1561630088. Same with a few other tweaks that solved regressions - see links in the PR description.

It wasn't *that* much worse, so I could bring it back if you think it's worth it but unfortunately the general pattern I see is that resolving these regressions on their own can often be detrimental, at least in the contracts we use for benchmarking.

Reply:
Can we instead change the test case, s.t. it optimizes well again :-D?

Reply:
These are artificial cases, so we shouldn't hand-tweak the optimizer for these over the external tests - but the test case is a bit weird - I mean, arguably the way it's now this is a *failing* test :-).

Reply:
So, I assume that the artificial part is the loop? I see it and the `if` above it were added in #5584 to test StructuralSimplifier. I removed it from the test and now the new sequence can deal with resolving the memory loads.

Still, I'm not sure if removing this is the right thing to do just because the new sequence can't handle it. It is a bit artificial, but it worked with the old sequence and the fact that it regressed is also potentially useful information.

I mean, maybe the case no longer makes sense with it (assuming it was meant to test memory load resolution - the name does not indicate that though), but then we could have a test specifically for the simplification:
```yul
{
    if sub(2, 1) {
        mstore(0x40, 0x20)
        for {} sub(1, 1) {} {}
    }
    sstore(1, mload(0x40))
}
```
I noticed that even this is a problem for the new sequence, even though it's much simpler. And to me it seems much less artificial. I could see some production code actually reducing to something close to this after passing through other steps. For now I added it as `fullSuite/structural_simplification.sol`.

Reply:
By the way, I noticed that StructuralSimplifier (specifically the one used in the `TOntnfDIul` segment) can deal with the loop if insert ExpressionSimplifier before it (`TOnstnfDIul`). And it makes sense because it looks like StructuralSimplifier on its own can't deal with even simple constant expressions, e.g. `for {} iszero(1) {} {}`. It needs the condition to be already resolved into `false`. Given that, shouldn't ExpressionSimplifier be recommended before StructuralSimplifier?

Also, ExpressionSimplifier only really helps it deal with the loop. The memory load is not resolved because simplification runs after the last occurrence of LoadResolver. I was curious if running the resolver once more, after the simplifier helps, and it does (when paired with CSE). So making it `TOnstcLnfDIul` lets the new sequence deal even with this artificial example. I did not add it to the sequence in the PR because it could again make things worse in other cases and would generally require rerunning all benchmarks. I can try it in a separate PR later though.